lab_2/main.py

```diff
@@ -119,26 +304,61 @@ def create_diff_report(original_text_tokens: tuple, suspicious_text_tokens: tupl
     """
-    pass
+    result = []
+    for sent_ind in range(len(original_text_tokens)):
+        # here prepare the sentence
+        original, suspicious = list(original_text_tokens[sent_ind]), list(suspicious_text_tokens[sent_ind])
+        differences = accumulated_diff_stats['difference_indexes']
+        number = 0
+        for index in differences[sent_ind][0]:
+            original.insert(index + number, '|')
+            suspicious.insert(index + number, '|')
+            number += 1
+        result.append(f"- {' '.join(original)}\n+ {' '.join(suspicious)}\n"
+                      f" \n"
+                      f"lcs = {accumulated_diff_stats['sentence_lcs_length'][sent_ind]}, "
+                      f"plagiarism = {accumulated_diff_stats['sentence_plagiarism'][sent_ind] * 100}%"
+                      f" \n")
+    result.append(f"Text average plagiarism (words): {accumulated_diff_stats['text_plagiarism'] * 100}%")
+    return "\n".join(result)
 
 
-def find_lcs_length_optimized(first_sentence_tokens: tuple, second_sentence_tokens: tuple,
-                              plagiarism_threshold: float) -> int:
+def find_lcs_length_optimized(first_sentence_tokens: list, second_sentence_tokens: list, plagiarism_threshold: float) -> int:
     """
-    Finds a length of the longest common subsequence using an optimized algorithm
-    When a length is less than the threshold, it becomes 0
-    :param first_sentence_tokens: a tuple of tokens
-    :param second_sentence_tokens: a tuple of tokens
-    :param plagiarism_threshold: a threshold
+    Finds a length of the longest common subsequence using the Hirschberg's algorithm
+    At the same time, if the first and last tokens coincide,
+    they are immediately added to lcs and not analyzed
+    :param first_sentence_tokens: a list of tokens
+    :param second_sentence_tokens: a list of tokens
     :return: a length of the longest common subsequence
     """
-    pass
+    # row that we create based on the previous
+    counted_row = [0] * len(second_sentence_tokens)
+
+    for row in range(len(first_sentence_tokens)):
+        # to update new row we create a copy (previous)
+        previous = counted_row.copy()
+        # the same logic as in usual, but save only the row
+        for column in range(len(second_sentence_tokens)):
+            if first_sentence_tokens[row] == second_sentence_tokens[column]:
+                counted_row[column] = previous[column - 1] + 1
+            else:
+                counted_row[column] = max(previous[column], counted_row[column - 1])
+    length = counted_row[-1]
+    if length / len(second_sentence_tokens) < plagiarism_threshold:
+        return 0
+
+    return length
 
 
 def tokenize_big_file(path_to_file: str) -> tuple:
-    """
-    Reads, tokenizes and transforms a big file into a numeric form
-    :param path_to_file: a path
-    :return: a tuple with ids
-    """
-    pass
+    with open("data.txt") as first_file, open("data_2.txt") as second_file:
```

Comment:
You do need to open these particular files. 
You should work with `path_to_file` argument.